### PR TITLE
Addison Cole - Chantelle Belic - RAILS RIDESHARE - C10 Edges

### PR DESCRIPTION
**Pull request by addisoncole**
# Rideshare-Rails
Congratulations! You're submitting your assignment! These comprehension questions should be answered by both partners together, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe the types of entity relationships you set up in your project and why you set up the relationships that way | Trips belong to drivers and passengers. Each trip has one passenger and one driver. Drivers or Passengers may have many trips. 
Describe the role of model validations in your application | They make sure that the user enters certain information deemed required or necessary. If a user creates a driver or passenger without the required fields or a trip that does not have the requirement of having a driver and a passenger then an error will be return to the user rather than HTML error code. 
How did your team break up the work to be done? | We pair programmed for the foundation of the app, such as building the database and model logic, creating the initial HTML view structure, the controller logic and routes. This way we talked through the initial structure together and made sure we shared a same vision for the set up. Once we got the initial structure we shifted to a more single programmer effort where we would each work on implementing different features that didnt overlap too much. Chantelle worked on validations largely, while Addison worked on creating the CSS styling. 
What features did you choose to prioritize in your project, and what features, if any, did you have to set aside to meet the deadline? | Some styling is not as organized as we would like, and displaying the model indexes as more formal tables took a backseat. Some other UX stuff such as rendering strings notifying user of save success.
What was one thing that your team collectively gained more clarity on after completing this assignment? |Where things/logic should be placed controller vs model vs view. When to use nested routes.
What is your Trello board URL? | https://trello.com/b/CIbLDshX
What is the Heroku URL of your deployed application? | https://rideshare-addy-chantelle-c10.herokuapp.com/home
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Getting better at explaining code or what code to type, working on better shared language of rails, we felt we were sometimes on different pages talking about the same thing because our shared language of the concepts was not as similar. Taking more time separately to whiteboard problems when we were stuck on a problem as a pair. Both working on git branching more.

### CODE REVIEW

**Comment by tildeee** [comment]
# Rideshare Rails
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate git usage with no extraneous files checked in and both partners contributing | x
Answered comprehension questions | x
Uses named routes (like `_path`) | x
RESTful routes utilized | x
**Project Requirements** |
Table relationships | x
Validation rules for Models | x
Business logic is in the models | x
Database is seeded from the CSV files | x
Trello board is created and utilized in project management | x
Heroku instance is online | x
The app is styled to create an attractive user interface | x
**Overall** |

Great work Addison and Chantelle! The project

In particular, I'm really impressed with the following bits of your project:
- The clean model logic that you two wrote
- How easy and delightful it was to use your app -- Not only does it look nice with styling, but the layout and usability in general was great. I could see all of the information I wanted to see in each page. Buttons changed names depending on which passenger or driver I was looking at. It was great!
- You two covered the interesting and subtle edge cases (even without tests)! You all remembered to check math when you may divide by 0 (the case if trips length was zero)

I'm leaving a few comments on your code, but good work overall! I'm really happy with this project.

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  root 'home#index'
+
+
+  resources :trips
+  resources :passengers
+   resources :passengers, except: :destroy  do
+    resources :trips, only: :create
+    member do
+      get :trips, to: 'passengers#index'
```
Comment (tildeee) [inline config/routes.rb:12]: Interesting! My understanding of this section of code is that it gives you a named path named `trips_passenger_path` for a GET at `/passengers/:id/trips` that will go to the PassengersController index action specifically. However, I don't see any use of the `trips_passenger_path` in this project... What did you use this member block for that I don't see?

File: app/models/trip.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  belongs_to :passenger
+
+  #helper method for create trip
+  def self.find_available_driver
+    drivers = Driver.all
+    driver = drivers.find_by(available?: true)
+    if driver
+      driver.update(available?: false)
+      driver.save
+      return driver
```
Comment (tildeee) [inline app/models/trip.rb:12]: Nice method! It has a lot of good logic in here. I've spoken to Chantelle about this, but it looks like the parts of logic that belong in `Trip` might remain as `find_available_driver`, but it might call some logic that lives in `Driver`... the bits about update and save in particular strike as "should belong in Driver", and may be called something like "become_unavailable"

File: app/models/driver.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    else
+      #perfect score if new
+      return 5
+    end
+  end
+
+  def find_available_driver
+    drivers = Driver.all
+    @driver = drivers.find_by(available?: true)
+    return @driver
```
Comment (tildeee) [inline app/models/driver.rb:35]: This method isn't being used! 1. If it's being unused, feel free to delete it. 2. Let's say it were being used -- you probably don't need to turn `@driver` into an instance variable.

File: app/models/driver.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      else
+        sum + trip.rating.to_f
+      end
+    end
+
+    if total_length != 0
+      return ("%.1f" % (total / total_length).to_f)
+    else
+      #perfect score if new
+      return 5
```
Comment (tildeee) [inline app/models/driver.rb:28]: Nice logic. It covers all cases :)

File: app/models/driver.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    total_length = self.trips.length
+
+    total = self.trips.reduce(0) do |sum, trip|
+      if trip.rating == nil
+        total_length -=1
+      else
+        sum + trip.rating.to_f
+      end
+    end
```
Comment (tildeee) [inline app/models/driver.rb:22]: nice use of `reduce`! It looks great!

File: app/models/driver.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
+class Driver < ApplicationRecord
+  has_many :trips, dependent: :destroy
+  validates :name, presence: true
+  validates :vin, presence: true
+
+  def total_revenue
+    return self.trips.reduce(0) do |sum, trip|
+      sum + ((trip.cost - 165)* 0.8)
+    end
```
Comment (tildeee) [inline app/models/driver.rb:9]: Nice!
